File: src/test/java/com/glitch/knowledge/KnowledgeInvalidRequestTests.java
package com.glitch.knowledge;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.web.servlet.MockMvc;

@SpringBootTest
@AutoConfigureMockMvc
@ActiveProfiles("test")
public class KnowledgeInvalidRequestTests {

	@Autowired
	private MockMvc mvc;

	@Test
	void testStepVerificationSameOriginDestination() throws Exception {
		// line 2, Avron -> Avron, direction Porte Dauphine
		mvc.perform(get("/maison/verify/line/2/origin/14/destination/14/direction/223"))
				.andExpect(status().is(400));
	}

	@Test
	void testStepVerificationInvalidLine() throws Exception {
		// line [invalid], Nation -> Avron, direction Porte Dauphine
		mvc.perform(get("/maison/verify/line/50/origin/189/destination/14/direction/223"))
				.andExpect(status().is(400));
	}

	@Test
	void testStepVerificationInvalidOrigin() throws Exception {
		// line 1, [invalid] -> Bastille, direction La Defense
		mvc.perform(get("/maison/verify/line/1/origin/500/destination/20/direction/135"))
				.andExpect(status().is(400));
	}

	@Test
	void testStepVerificationInvalidDestination() throws Exception {
		// line 1, Gare De Lyon -> [Invalid], direction La Defense
		mvc.perform(get("/maison/verify/line/1/origin/111/destination/500/direction/135"))
				.andExpect(status().is(400));
	}

	@Test
	void testStepVerificationInvalidDirection() throws Exception {
		// line 1, Gare De Lyon -> Bastille, direction [Invalid]
		mvc.perform(get("/maison/verify/line/1/origin/111/destination/20/direction/500"))
				.andExpect(status().is(400));
	}

	@Test
	void testStepVerificationDirectionNotConnectedToLine() throws Exception {
		// line 1, Gare De Lyon -> Bastille, direction Porte Dauphine[wrong]
		mvc.perform(get("/maison/verify/line/1/origin/111/destination/20/direction/223"))
				.andExpect(status().is(400));
	}

}

File: src/test/kotlin/com/glitch/knowledge/KnowledgeInvalidRequestTests.kt
package com.glitch.knowledge

import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.web.servlet.MockMvc


@SpringBootTest
@AutoConfigureMockMvc
@ActiveProfiles("test")
class KnowledgeInvalidRequestTests {
    @field:Autowired
    lateinit var mvc: MockMvc

    var stationMap: Map<String, Int> = mapOf()
    var lineMap: Map<String, Int> = mapOf()

    init {
        stationMap = readStationListJson()
        lineMap = lineMap()
    }

    @Test
    fun `invalid request because origin = destination`() {
        runVerificationTestInvalid(
            mvc,
            lineMap["2"],
            stationMap["Avron"],
            stationMap["Avron"],
            stationMap["Porte Dauphine"],
        )
    }

    @Test
    fun `invalid request, the line id does not exist`() {
        runVerificationTestInvalid(
            mvc,
            500,
            stationMap["Nation"],
            stationMap["Avron"],
            stationMap["La Défense"],
        )
    }

    @Test
    fun `invalid request, the origin id does not exist`() {
        runVerificationTestInvalid(
            mvc,
            lineMap["1"],
            500,
            stationMap["Bastille"],
            stationMap["La Défense"],
        )
    }

    @Test
    fun `invalid request, the destination id does not exist`() {
        runVerificationTestInvalid(
            mvc,
            lineMap["1"],
            stationMap["Gare de Lyon"],
            500,
            stationMap["La Défense"],
        )
    }

    @Test
    fun `invalid request, the direction id does not exist`() {
        runVerificationTestInvalid(
            mvc,
            lineMap["1"],
            stationMap["Gare de Lyon"],
            stationMap["Bastille"],
            500,
        )
    }

    @Test
    fun `invalid request, the direction is not a terminus of the line`() {
        runVerificationTestInvalid(
            mvc,
            lineMap["1"],
            stationMap["Gare de Lyon"],
            stationMap["Bastille"],
            stationMap["Porte Dauphine"],
        )
    }
}